docs/static/dead-letter-queues.asciidoc

```diff
@@ -0,0 +1,149 @@
+[[dead-letter-queues]]
+=== Dead Letter Queues
+
+NOTE: The dead letter queue feature is currently supported for the
+<<plugins-outputs-elasticsearch>> output only. Support for additional outputs
+will be available in future releases of the Logstash plugins. Before configuring
+Logstash to use this feature, refer to the output plugin documentation to
+verify that the plugin supports the dead letter queue feature.
+
+By default, when Logstash encounters an event that it cannot process because the
+data contains a mapping error or some other issue, the Logstash pipeline 
+either hangs or drops the unsuccessful event. In order to protect against data
+loss in this situation, you can configure Logstash to write unsuccessful events
+to a dead letter queue instead of dropping them. 
+
+Each event written to the dead letter queue includes the original event along
+with metadata indicating when the event entered the queue. For example:
+
+//TODO: Need a better example here. 
+
+[source,json]
+-------------------------------------------------------------------------------
```

Comment:
@dedemorton - is this helpful? The reason field in the object is pretty printed to a 'realistic' level - the reason value is written as a string rather than structured data, which makes it the embedded JSON a little uglier.

Example scenario: 

A user attempts to index a document including geo_ip data using the logstash-output-elasticsearch plugin, but the data
cannot be processed due to a mapping error: 

eg - 
```
{"geoip":{"location":"home"}}
```
Would fail to ingest with a mapping error, as the logstash output plugin expects the location field to be a geo_point object.

This failure would be written to the dead letter queue as follows:

```
{
  "geoip" : {
    "location" : "home"
  },
  "@timestamp" : "2017-06-21T18:07:36.434Z",
  "@metadata" : {
    "dead_letter_queue" : {
      "plugin_id" : "4e2feb09821f03a96ed88786e2f6c248b46d4f3b-2",
      "reason" : "Could not index event to Elasticsearch. status: 400, action: [\"index\", {:_id=>nil, :_index=>\"logstash-2017.06.21\", :_type=>\"logs\", :_routing=>nil}, #<LogStash::Event:0x177905e6>], response: {\"index\"=>{\"_index\"=>\"logstash-2017.06.21\", \"_type\"=>\"logs\", \"_id\"=>\"AVzL1pP_jH06PKfP7bY6\", \"status\"=>400, \"error\"=>{\"type\"=>\"mapper_parsing_exception\", \"reason\"=>\"failed to parse\", \"caused_by\"=>{\"type\"=>\"illegal_argument_exception\", \"reason\"=>\"illegal latitude value [266.30859375] for geoip.location\"}}}}",
      "entry_time" : "2017-06-21T18:07:36.463Z",
      "plugin_type" : "elasticsearch"
    }
  },
  "@version" : "1",
  "host" : "Robs-MacBook-Pro.local"
}
```
The original data is included as provided, with additional metadata written by the logstash-output-elasticsearch plugin,
which states the reason why the event could not be processed by the plugin, along with information relating to the plugin and
the time that the event as entered into the dead letter queue. 

From here, the event can be processed by the logstash-input-dead_letter_queue and transformed into a form that can successfully
be processed by the logstash-output-elasticsearch plugin.

For example, the following pipeline configuration would remove the offending field from the document, and allow it to be
processed:
```
input { 
  dead_letter_queue { 
    path => "/path.to/data/dead_letter_queue/" 
  } 
} 
filter {
  json {
    source => [original] 
  } 
  mutate { 
    remove_field => "[geoip][location]" 
  } 
} 
output { 
  elasticsearch{} 
}
```